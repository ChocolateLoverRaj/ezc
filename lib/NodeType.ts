enum NodeType {
  VARIABLE_INITIALIZE,
  FUNCTION_CALL,
  FUNCTION,
  RETURN,
  NUMBER_LITERAL,
  INPUT
}

export default NodeType
