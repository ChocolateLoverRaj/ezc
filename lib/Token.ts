enum Token {
  // <
  RETURN,

  // ;
  SEMICOLON,

  // int goodNumber = 3
  INT,

  // 3, 23423, 93
  NUMBER_LITERAL,

  // byte b = 255
  BYTE,

  // "Hello world!"
  STRING_LITERAL,

  // (
  PARENTHESIS_OPEN,
  // )
  PARENTHESIS_CLOSE,

  // auto a = () > 234
  AUTO,

  // myFunction()
  IDENTIFIER
}

export default Token;
