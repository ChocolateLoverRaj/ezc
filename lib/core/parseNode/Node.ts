import EnumItem from '../EnumItem'

interface Node {
  type: EnumItem
  data: any
}

export default Node
