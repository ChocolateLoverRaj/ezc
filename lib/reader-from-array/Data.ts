interface Data<T> {
  array: readonly T[]
  index: number
}

export default Data
