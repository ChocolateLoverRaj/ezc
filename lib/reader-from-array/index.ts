export { default as Data } from './Data'
export { default as getReader } from './getReader'
export { default as initialize } from './initialize'
