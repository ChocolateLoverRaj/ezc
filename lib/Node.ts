import NodeType from './NodeType'
import TypeToken from './TypeToken'

interface NodeDataMappings {
  [NodeType.NUMBER_LITERAL]: number
  [NodeType.RETURN]: Node
  [NodeType.VARIABLE_INITIALIZE]: {
    /**
     * Is it `int`, `double`, `float`, `auto`?
     */
    type: TypeToken
    /**
     * Example code: `int n`. `n` is this
     */
    variableIdentifier: string
    /**
     * Example: `int n = 1`. `1` is this
     */
    initializedValue?: Node
  }
}

type HiddenInterface = {
  [K in keyof NodeDataMappings]: {
    type: K
    data: NodeDataMappings[K]
  };
} &
{
  [K in NodeType]: {
    type: K
  };
}

type Node = HiddenInterface[keyof HiddenInterface]

export default Node
