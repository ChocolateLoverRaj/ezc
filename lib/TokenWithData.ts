import Token from './Token'

interface TokenDataMapping {
  [Token.NUMBER_LITERAL]: number
  [Token.STRING_LITERAL]: string
  [Token.IDENTIFIER]: string
}

type HiddenInterface = {
  [K in keyof TokenDataMapping]: {
    token: K
    data: TokenDataMapping[K]
  };
} &
{
  [K in Token]: {
    token: K
  };
}

type TokenWithData = HiddenInterface[keyof HiddenInterface]

export default TokenWithData
