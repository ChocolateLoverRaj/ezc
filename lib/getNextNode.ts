import Reader from './Reader'
import TokenWithData from './TokenWithData'
import Node from './Node'
import Token from './Token'
import NodeType from './NodeType'
import never from 'never'
import typeTokens from './typeTokens'
import TypeToken from './TypeToken'
import NodeByType from './NodeByType'

const getNextNode = (
  reader: Reader<TokenWithData>,
  mustBeExpression = false
): Node => {
  const {
    getCurrent,
    next
  } = reader
  const tokenWithData = getCurrent() ?? never('No token')
  if (!mustBeExpression && tokenWithData.token === Token.ARROW_LEFT) {
    next()
    return {
      type: NodeType.RETURN,
      data: getNextNode(reader, true)
    }
  }
  if (tokenWithData.token === Token.NUMBER_LITERAL) {
    next()
    return {
      type: NodeType.NUMBER_LITERAL,
      data: tokenWithData.data
    }
  }
  if (typeTokens.includes(tokenWithData.token)) {
    next()
    const identifierToken = getCurrent()
    if (identifierToken?.token !== Token.IDENTIFIER) {
      throw new Error('No identifier after type')
    }
    next()
    const initializedValue = (() => {
      const assignmentToken = getCurrent()
      if (assignmentToken?.token === Token.ASSIGNMENT) {
        next()
        return getNextNode(reader, true)
      }
    })()
    return {
      type: NodeType.VARIABLE_INITIALIZE,
      data: {
        type: tokenWithData.token as TypeToken,
        variableIdentifier: identifierToken.data,
        initializedValue
      }
    }
  }
  if (tokenWithData.token === Token.PARENTHESIS_OPEN) {
    next()
    const nextToken = getCurrent() ?? never('No token after (')
    if (typeTokens.includes(nextToken.token) || nextToken.token === Token.PARENTHESIS_CLOSE) {
      const inputs: Array<NodeByType[NodeType.INPUT]> = []
      if (nextToken.token === Token.PARENTHESIS_CLOSE) {
        next()
      } else {
        while (true) {
          const typeToken = getCurrent() ?? never('No token after start of input list')
          next()
          if (!typeTokens.includes(typeToken.token)) {
            throw new Error('Not type')
          }
          const identifierToken = getCurrent()
          next()
          if (identifierToken?.token !== Token.IDENTIFIER) {
            throw new Error('Not identifier')
          }
          inputs.push({
            type: NodeType.INPUT,
            data: {
              type: typeToken.token,
              identifier: identifierToken.data
            }
          })
          const nextToken = getCurrent() ?? never('Unclosed input list')
          next()
          if (nextToken.token === Token.PARENTHESIS_CLOSE) break
          if (nextToken.token !== Token.COMMA) {
            throw new Error('Bad token')
          }
        }
      }
      {
        const nextToken = getCurrent() ?? never('No token after input list')
        next()
        if (nextToken.token !== Token.ARROW_RIGHT) {
          throw new Error('Not > after input list')
        }
      }
      {
        const nextToken = getCurrent() ?? never('No token after > after input list')
        if (nextToken.token === Token.CURLY_BRACKET_OPEN) {
          throw new Error('Curly bracket function inside not supported yet')
        }
      }
      return {
        type: NodeType.FUNCTION,
        data: {
          inputs,
          inside: [{
            type: NodeType.RETURN,
            data: getNextNode(reader, true)
          }]
        }
      }
    }
    return getNextNode(reader, true)
  }
  throw new Error('Invalid token')
}

export default getNextNode
