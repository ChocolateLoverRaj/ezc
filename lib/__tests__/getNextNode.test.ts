import getNextNode from '../getNextNode'
import Node from '../Node'
import NodeType from '../NodeType'
import { getReader, initialize } from '../reader-from-array'
import Token from '../Token'
import TokenWithData from '../TokenWithData'

test('< 3', () => {
  const data = initialize<TokenWithData>([{
    token: Token.ARROW_LEFT
  }, {
    token: Token.NUMBER_LITERAL,
    data: 3
  }])
  const node = getNextNode(getReader(data))
  expect(node).toStrictEqual<Node>({
    type: NodeType.RETURN,
    data: {
      type: NodeType.NUMBER_LITERAL,
      data: 3
    }
  })
  expect(data.index).toBe(2)
})

test('<', () => {
  const data = initialize<TokenWithData>([{
    token: Token.ARROW_LEFT
  }])
  expect(() => getNextNode(getReader(data))).toThrowErrorMatchingSnapshot()
})

test('< <', () => {
  const data = initialize<TokenWithData>([{
    token: Token.ARROW_LEFT
  }, {
    token: Token.ARROW_LEFT
  }])
  expect(() => getNextNode(getReader(data))).toThrowErrorMatchingSnapshot()
})

test('int32 n', () => {
  const data = initialize<TokenWithData>([{
    token: Token.INT32
  }, {
    token: Token.IDENTIFIER,
    data: 'n'
  }])
  const node = getNextNode(getReader(data))
  expect(node).toStrictEqual<Node>({
    type: NodeType.VARIABLE_INITIALIZE,
    data: {
      type: Token.INT32,
      variableIdentifier: 'n',
      initializedValue: undefined
    }
  })
  expect(data.index).toBe(2)
})

test('int32 n = 1', () => {
  const data = initialize<TokenWithData>([{
    token: Token.INT32
  }, {
    token: Token.IDENTIFIER,
    data: 'n'
  }, {
    token: Token.ASSIGNMENT
  }, {
    token: Token.NUMBER_LITERAL,
    data: 1
  }])
  const node = getNextNode(getReader(data))
  expect(node).toStrictEqual<Node>({
    type: NodeType.VARIABLE_INITIALIZE,
    data: {
      type: Token.INT32,
      variableIdentifier: 'n',
      initializedValue: {
        type: NodeType.NUMBER_LITERAL,
        data: 1
      }
    }
  })
  expect(data.index).toBe(4)
})

test('() > 234', () => {
  const data = initialize<TokenWithData>([{
    token: Token.PARENTHESIS_OPEN
  }, {
    token: Token.PARENTHESIS_CLOSE
  }, {
    token: Token.ARROW_RIGHT
  }, {
    token: Token.NUMBER_LITERAL,
    data: 234
  }])
  const node = getNextNode(getReader(data))
  expect(node).toStrictEqual<Node>({
    type: NodeType.FUNCTION,
    data: {
      inputs: [],
      inside: [{
        type: NodeType.RETURN,
        data: {
          type: NodeType.NUMBER_LITERAL,
          data: 234
        }
      }]
    }
  })
  expect(data.index).toBe(4)
})

test.todo('() > { return 3 }')
