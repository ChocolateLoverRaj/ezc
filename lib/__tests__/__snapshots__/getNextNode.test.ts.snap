// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`< < 1`] = `"Invalid token"`;

exports[`< 1`] = `"No token"`;
